lib/configuration.js

```diff
@@ -50,150 +54,225 @@ function applyLoggingOptions(options) {
+  R.dissocPath(['options', 'level']),
+  R.dissocPath(['options', 'q']),
+  R.dissocPath(['options', 'silent']),
+  R.dissocPath(['options', 't']),
+  R.dissocPath(['options', 'timestamp']),
+  R.dissocPath(['options', 'blueprintPath']),
+  R.dissoc('data')
+);
+
+// TODO Finish and test keys
+const coerceOptions = R.evolve({
+  apiDescriptions: R.compose(
+    R.map((content, index) => ({
+      location: `configuration.apiDescriptions[${index}]`,
+      content,
+    })),
+    coerceToArray
+  ),
+  reporter: coerceToArray,
+  output: coerceToArray,
+  header: coerceToArray,
+  method: R.compose(R.map(R.toUpper), coerceToArray),
+  only: coerceToArray,
+  path: coerceToArray,
+  hookfiles: coerceToArray,
+});
+
+// TODO Still confusing how to adequately compose { errors, warnings }
+const deprecationWarnings = {
+  level: 'Do not use `c`',
+};
 
-function coerceRemovedOptions(config = {}) {
-  const errors = [];
-  const warnings = [];
+const deprecationErrors = {
+  level: 'REMOVED level',
+  timestamp: 'REMOVE timestamp',
+};
 
-  if (config.options) {
-    if (typeof config.options.color === 'string') {
-      config.options.color = coerceToBoolean(config.options.color);
-    }
-    if (config.options.level || config.options.l) {
-      const level = config.options.level || config.options.l;
-      let loglevel;
-      if (['silly', 'debug', 'verbose'].includes(level)) {
-        loglevel = 'debug';
-      } else if (level === 'error') {
-        loglevel = 'error';
-      } else if (level === 'silent') {
-        loglevel = 'silent';
-      } else {
-        loglevel = 'warn';
-      }
-      config.options.loglevel = loglevel;
-      config.options.l = loglevel;
-      delete config.options.level;
-    }
-  }
-  if (config.data) {
-    warnings.push("DEPRECATED: The 'data' configuration property is deprecated "
-      + "in favor of 'apiDescriptions', please see https://dredd.org");
-
-    const apiDescriptions = Object.keys(config.data).map((location) => {
-      if (typeof config.data[location] === 'string') {
-        return {
-          location,
-          content: config.data[location],
-        };
-      }
-      return {
-        location: config.data[location].filename,
-        content: config.data[location].raw,
-      };
-    });
-    config.apiDescriptions = config.apiDescriptions
-      ? config.apiDescriptions.concat(apiDescriptions)
-      : apiDescriptions;
-    delete config.data;
+function flushMessages(rules, config) {
+  return Object.keys(rules).reduce((messages, optionName) => {
+    return config[optionName]
+      ? messages.concat(rules[optionName])
+      : messages;
+  }, []);
+}
+
+/**
+ * Returns the errors and warnings relative to the given config.
+ */
+const validateConfig = config => ({
+  errors: flushMessages(deprecationErrors, config),
+  warnings: flushMessages(deprecationWarnings, config),
+});
+
+const sanitizeConfig = R.compose(
+  coerceOptions,
+  removeDeprecatedOptions
```

Comment:
I think the best behavior would be:

- If people run Dredd with an unknown option, Dredd fails fast with an error this is unknown configuration.
- If people run Dredd with a previously known but now deprecated option, Dredd fails fast with an error this option has been removed.
- If the configuration code can figure out that a provided value is invalid for its respective option (such as `--loglevel` where it's a limited enum of available levels or `--color/--no-color`, which should be boolean), it should fail fast with error it's an invalid option.
- As a result of the above point, if an option hasn't been removed (like `-l`), but it's semantics had changed, Dredd should be able to fail fast for invalid (deprecated) values.
- If figuring out the validity of the option is not cheap (e.g. figuring out whether a file exists), it is passed along and left up to subsequent Dredd code.

The question is whether we want to implement all that in this PR or whether it's fine to do just a part of it now.

Reply:
Fail fast approach is very applicable here. I was thinking about that, but I would first rewrite it to the state that would resemble how it worked previously, and then it's up to us to decide whether to pile up this pull request, or merge and deal with that in the next one.

Reply:
Would it make sense to do it like this then?

1. PR for refactoring the config
2. PR for dropping Node 6 support (= breaking change) & demoting most of the softly deprecated options (errors) to being removed (warnings) and failing fast (= breaking change)

Reply:
Yes, absolutely. Let me revert the nodeJS deprecation and focus on configuration composition at the moment.